shell/server/installer.js

```diff
@@ -70,28 +84,40 @@ deletePackage = function (packageId) {
   }
 }
 
-startInstall = function (packageId, url, fromBeginning, appId) {
-  if (packageId in installers) {
-    return;
+startInstall = function (packageId, url, retryFailed) {
+  // Mark package for possible installation.
+
+  var fields = {status: "download", progress: 0, url: url};
+
+  if (retryFailed) {
+    Packages.update({_id: packageId, status: "failed"}, {$set: fields});
+  } else {
+    fields._id = packageId;
+    Packages.insert(fields);
```

Comment:
It looks like this could throw an exception if two people try to install the same package at approximately the same time. It's an unlikely case, and this is not a catastrophic way to handle it, but perhaps there is a better way?